libethereum/EthereumHost.cpp

```diff
@@ -576,8 +576,20 @@ void EthereumHost::onPeerNewBlock(EthereumPeer* _peer, RLP const& _r)
 				u256 difficulty = _r[1].toInt<u256>();
 				if (m_syncingTotalDifficulty < difficulty)
 				{
+					BlockInfo bi;
+					RLP root(_r[0].data());
+					try
+					{
+						bi.populateFromHeader(root[0], CheckEverything, h);
```

Comment:
If an attacker did that it wouldn't get to this line. The difficulty of the block is checked on line 577 before bothering with header/nonce check.
It doesn't matter what the BlockQueue does. Lines 591+ irrevocably modify state prior to *any* verification of the block or difficulty. The block could be filled with 0s except for the difficulty and without checking the header/nonce it will immediately begin syncing to the peer. Would it not?

I understand there maybe a PR in the future that will tag the block with the peer? I think until then, this PR will prevent certain malformed blocks from fooling the client into sync'ing to the peer.